### PR TITLE
Add Lexer

### PR DESCRIPTION
**Pull request by NeilKleistGao**
Related Issue: #1

This PR is still not ready.

### CODE REVIEW

File: src/main/scala/arcana/empress/Lexer.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      }
+      else if (str(column) == '"') parseStringLiteral(str, Position(line, column)) match {
+        case (locatedToken, next) => parseLine(parsed :+ locatedToken, next)
+      }
+      else if (str(column).isDigit) parseNumericLiteral(str, Position(line, column)) match {
+        case (locatedToken, next) => parseLine(parsed :+ locatedToken, next)
+      }
+      else parseIdentifierAndKeyword(str, Position(line, column)) match {
+        case (locatedToken, next) => parseLine(parsed :+ locatedToken, next)
+      }
```
Comment (Cheny-chui) [inline src/main/scala/arcana/empress/Lexer.scala:30]: why use three match here

File: src/main/scala/arcana/empress/Lexer.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+object Lexer {
+  def apply(code: String, options: LexerOptions) =
+    new Lexer(code, options)
+
+  private val keywords = List[String](
+    "def", "defn", "let", "if", "match" // TODO: Add more if need
+  )
+
+  private val punctuations = List[Char](
+    '(', ')', ' ', '\t', '\n', '\r', ':', '[', ']', '{', '}' // TODO: Are these symbols enough?
```
Comment (Cheny-chui) [inline src/main/scala/arcana/empress/Lexer.scala:88]: do +-*/ these mathematical symbols belongs to Lexer‘s duty？

File: src/main/scala/arcana/empress/Lexer.scala (hunk 1 above)
Comment (NeilKleistGao) [reply src/main/scala/arcana/empress/Lexer.scala:30]: Deconstruct the pair result without creating a temp variable.

File: src/main/scala/arcana/empress/Lexer.scala (hunk 2 above)
Comment (NeilKleistGao) [reply src/main/scala/arcana/empress/Lexer.scala:88]: I treat symbols `+-*/` (and so on) like identifiers. Then we can allow users to define their own operators.

File: src/main/scala/arcana/empress/Lexer.scala (hunk 1 above)
Comment (Cheny-chui) [reply src/main/scala/arcana/empress/Lexer.scala:30]: but you always return same pair result, way just follow with parseLine without match

Comment (NeilKleistGao) [reply src/main/scala/arcana/empress/Lexer.scala:30]: What?

**Review by Cheny-chui — COMMENTED** [review]
LGTM

**Review by Cheny-chui — APPROVED** [review]
apporve merge

**Pull request merged**